File: hiob_contracts/factory/planet_output.py
"""`PlanetOutput` and `ArtifactRef` — the immutable envelopes every Planet emits.

PRD_CREATIVE_FACTORY_HARMONY §6.1 (`PlanetOutputV1<T>`) and §6.4 (`ArtifactRefV1`).

A Planet never hands another Planet a bare payload. It emits a `PlanetOutput`:
an immutable, digest-linked envelope carrying the payload as a canonical JSON
object plus references to any binary artifacts. Binary bytes never travel in the
JSON edge envelope — only `ArtifactRef`s that point at them.

The payload's *type* is owned by the producing Planet and `hiob-contracts`
(referenced via `ContractRef.schema_digest`), not by this envelope; the envelope
carries the already-validated payload as a JSON object and binds it by digest.
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field, model_validator

from .digest import Digest, assert_digest, is_digest, sha256_digest

_FROZEN = {"frozen": True, "extra": "forbid"}


class ContractRef(BaseModel):
    """Reference to a versioned contract schema (name + version + schema digest)."""

    model_config = _FROZEN

    name: str
    version: str
    schema_digest: Digest

    @model_validator(mode="after")
    def _check_digest(self) -> "ContractRef":
        assert_digest(self.schema_digest, "contract.schema_digest")
        return self


class ArtifactRef(BaseModel):
    """Immutable reference to one produced binary artifact (§6.4).

    JSON carries this reference; the bytes live at `uri` and are bound by `sha256`.
    Lineage is explicit: `source_output_digests` and `edge_receipt_digests` record
    which upstream outputs and Karma edges produced this artifact.
    """

    model_config = _FROZEN

    artifact_id: str
    kind: str
    uri: str
    sha256: Digest
    mime: str
    bytes_len: int = Field(ge=0)
    duration_ms: int | None = Field(default=None, ge=0)
    width: int | None = Field(default=None, ge=0)
    height: int | None = Field(default=None, ge=0)
    beat_index: int | None = Field(default=None, ge=0)
    producer_planet: str
    producer_node_id: str
    execution_id: str
    producer_revision: str
    image_digest: Digest | None = None
    source_output_digests: tuple[Digest, ...] = ()
    edge_receipt_digests: tuple[Digest, ...] = ()
    provenance_refs: tuple[str, ...] = ()
    consent_refs: tuple[str, ...] = ()

    @model_validator(mode="after")
    def _check_digests(self) -> "ArtifactRef":
        assert_digest(self.sha256, "artifact.sha256")
        if self.image_digest is not None:
            assert_digest(self.image_digest, "artifact.image_digest")
        for d in (*self.source_output_digests, *self.edge_receipt_digests):
            if not is_digest(d):
                raise ValueError(f"artifact lineage digest malformed: {d!r}")
        return self


class PlanetOutput(BaseModel):
    """Immutable typed output a Planet returns over HTTP (§6.1).

    `payload` is a canonical JSON object already validated against `contract`.
    `payload_digest` binds the payload; `output_digest` binds the whole envelope
    (every field except `output_digest` itself). Use `finalize()` to compute both
    deterministically and `verify()` to reject tampering downstream.
    """

    model_config = _FROZEN

    schema_: str = Field(default="hiob.planet-output.v1", alias="schema")
    output_id: str
    run_id: str
    factory_revision: int = Field(ge=0)
    workspace_id: str
    trace_id: str
    execution_id: str
    attempt_no: int = Field(ge=1)
    producer: dict[str, str]  # {planet, node_id, revision}
    contract: ContractRef
    payload: dict[str, Any]
    payload_digest: Digest
    artifacts: tuple[ArtifactRef, ...] = ()
    output_digest: Digest
    prior_edge_receipt_id: str | None = None
    emitted_at: str

    @model_validator(mode="after")
    def _check(self) -> "PlanetOutput":
        for key in ("planet", "node_id", "revision"):
            if key not in self.producer:
                raise ValueError(f"producer missing required key: {key}")
        if self.payload_digest != sha256_digest(self.payload):
            raise ValueError("payload_digest does not match payload")
        if self.output_digest != _compute_output_digest(self):
            raise ValueError("output_digest does not match envelope contents")
        return self

    @classmethod
    def build(
        cls,
        *,
        output_id: str,
        run_id: str,
        factory_revision: int,
        workspace_id: str,
        trace_id: str,
        execution_id: str,
        attempt_no: int,
        producer: dict[str, str],
        contract: ContractRef,
        payload: dict[str, Any],
        emitted_at: str,
        artifacts: tuple[ArtifactRef, ...] = (),
        prior_edge_receipt_id: str | None = None,
    ) -> "PlanetOutput":
        """Construct with `payload_digest`/`output_digest` computed deterministically."""
        payload_digest = sha256_digest(payload)
        draft = cls.model_construct(
            schema_="hiob.planet-output.v1",
            output_id=output_id,
            run_id=run_id,
            factory_revision=factory_revision,
            workspace_id=workspace_id,
            trace_id=trace_id,
            execution_id=execution_id,
            attempt_no=attempt_no,
            producer=dict(producer),
            contract=contract,
            payload=dict(payload),
            payload_digest=payload_digest,
            artifacts=tuple(artifacts),
            output_digest="sha256:" + "0" * 64,
            prior_edge_receipt_id=prior_edge_receipt_id,
            emitted_at=emitted_at,
        )
        output_digest = _compute_output_digest(draft)
        # Re-validate through the normal path so illegal states still fail closed.
        return cls.model_validate(
            {**draft.model_dump(by_alias=True), "output_digest": output_digest}
        )

    def verify(self) -> bool:
        """Recompute digests and confirm the envelope is internally consistent."""
        return self.payload_digest == sha256_digest(self.payload) and (
            self.output_digest == _compute_output_digest(self)
        )


def _compute_output_digest(out: PlanetOutput) -> Digest:
    """Digest of the whole envelope with `output_digest` excluded (avoids self-ref)."""
    body = out.model_dump(by_alias=True)
    body.pop("output_digest", None)
    return sha256_digest(body)

File: ts/src/factory/planet-output.ts
/**
 * PlanetOutput + ArtifactRef — TS mirror of factory/planet_output.py.
 *
 * ⚠️ AUTHORITY: Python. envelope 다이제스트는 Python `model_dump(by_alias=True)`에서
 * `output_digest`를 제거한 canonical-JSON SHA-256과 일치해야 한다.
 */
import { z } from 'zod';
import { DIGEST_RE, Digest, sha256Digest } from './digest.js';

export const DigestSchema = z.string().regex(DIGEST_RE);

export const ContractRefSchema = z
  .object({
    name: z.string(),
    version: z.string(),
    schema_digest: DigestSchema,
  })
  .strict();
export type ContractRef = z.infer<typeof ContractRefSchema>;

export const ArtifactRefSchema = z
  .object({
    artifact_id: z.string(),
    kind: z.string(),
    uri: z.string(),
    sha256: DigestSchema,
    mime: z.string(),
    bytes_len: z.number().int().nonnegative(),
    duration_ms: z.number().int().nonnegative().nullable().default(null),
    width: z.number().int().nonnegative().nullable().default(null),
    height: z.number().int().nonnegative().nullable().default(null),
    beat_index: z.number().int().nonnegative().nullable().default(null),
    producer_planet: z.string(),
    producer_node_id: z.string(),
    execution_id: z.string(),
    producer_revision: z.string(),
    image_digest: DigestSchema.nullable().default(null),
    source_output_digests: z.array(DigestSchema).default([]),
    edge_receipt_digests: z.array(DigestSchema).default([]),
    provenance_refs: z.array(z.string()).default([]),
    consent_refs: z.array(z.string()).default([]),
  })
  .strict();
export type ArtifactRef = z.infer<typeof ArtifactRefSchema>;

const ProducerSchema = z
  .object({ planet: z.string(), node_id: z.string(), revision: z.string() })
  .strict();

export const PlanetOutputSchema = z
  .object({
    schema: z.literal('hiob.planet-output.v1').default('hiob.planet-output.v1'),
    output_id: z.string(),
    run_id: z.string(),
    factory_revision: z.number().int().nonnegative(),
    workspace_id: z.string(),
    trace_id: z.string(),
    execution_id: z.string(),
    attempt_no: z.number().int().min(1),
    producer: ProducerSchema,
    contract: ContractRefSchema,
    payload: z.record(z.string(), z.unknown()),
    payload_digest: DigestSchema,
    artifacts: z.array(ArtifactRefSchema).default([]),
    output_digest: DigestSchema,
    prior_edge_receipt_id: z.string().nullable().default(null),
    emitted_at: z.string(),
  })
  .strict()
  .superRefine((o, ctx) => {
    if (o.payload_digest !== sha256Digest(o.payload)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'payload_digest does not match payload' });
    }
    if (o.output_digest !== computeOutputDigest(o)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'output_digest does not match envelope' });
    }
  });
export type PlanetOutput = z.infer<typeof PlanetOutputSchema>;

/** envelope에서 `output_digest`를 제외한 canonical-JSON SHA-256 (Python `_compute_output_digest`). */
export function computeOutputDigest(out: Record<string, unknown>): Digest {
  const body: Record<string, unknown> = { ...out };
  delete body.output_digest;
  return sha256Digest(body);
}

export interface PlanetOutputInput {
  output_id: string;
  run_id: string;
  factory_revision: number;
  workspace_id: string;
  trace_id: string;
  execution_id: string;
  attempt_no: number;
  producer: { planet: string; node_id: string; revision: string };
  contract: ContractRef;
  payload: Record<string, unknown>;
  emitted_at: string;
  artifacts?: ArtifactRef[];
  prior_edge_receipt_id?: string | null;
}

/** payload_digest/output_digest를 결정적으로 채워 PlanetOutput 생성 (Python `build`). */
export function buildPlanetOutput(input: PlanetOutputInput): PlanetOutput {
  const artifacts = (input.artifacts ?? []).map((a) => ArtifactRefSchema.parse(a));
  const payload_digest = sha256Digest(input.payload);
  const envelope = {
    schema: 'hiob.planet-output.v1' as const,
    output_id: input.output_id,
    run_id: input.run_id,
    factory_revision: input.factory_revision,
    workspace_id: input.workspace_id,
    trace_id: input.trace_id,
    execution_id: input.execution_id,
    attempt_no: input.attempt_no,
    producer: input.producer,
    contract: input.contract,
    payload: input.payload,
    payload_digest,
    artifacts,
    prior_edge_receipt_id: input.prior_edge_receipt_id ?? null,
    emitted_at: input.emitted_at,
  };
  const output_digest = computeOutputDigest(envelope);
  return PlanetOutputSchema.parse({ ...envelope, output_digest });
}

/** 다이제스트 재계산으로 envelope 내부 정합성 확인 (Python `verify`). */
export function verifyPlanetOutput(out: PlanetOutput): boolean {
  return out.payload_digest === sha256Digest(out.payload) && out.output_digest === computeOutputDigest(out);
}
